File: pe/pe15.rb
=begin
LEN = 20
c = 0
q = []
q.push [0, 0]
loop do
  x, y = q.pop
  break if x.nil? && y.nil?
  if x == LEN && y == LEN
    c += 1
  else
    q.push [x + 1, y] if x < LEN
    q.push [x, y + 1] if y < LEN
  end
end
p c
=end

len = 20 + 1
map = Array.new(len) {Array.new(len) {0}}
len.times{|i| map[0][i] = 1}
len.times{|i| map[i][0] = 1}
(1...len).each{|y|
  (1...len).each{|x| map[y][x] = map[y - 1][x] + map[y][x - 1]}
}
p map[len - 1][len - 1]

File: pe/pe15.ml
(*
let walk len =
  let rec _w x y t =
    if x = len && y = len then t + 1
    else (
      let t = if x < len then _w (x + 1) y t else t in
      if y < len && not (x = 0 && y = 0) then _w x (y + 1) t else t
    )
  in
  (_w 0 0 0) * 2

let walk len =
  let c = ref 0 in
  let s = Stack.create () in
  Stack.push (0, 0) s;
  try 
    while true do
      let x, y = Stack.pop s in
      if x = len && y = len then c := !c + 1
      else (
        if x < len then Stack.push (x + 1, y) s;
        if y < len then Stack.push (x, y + 1) s
      )
    done;
    !c
  with Stack.Empty -> !c

let _ =
  let rec loop i n =
    if i > n then ()
    else (
      Printf.printf "i(%d) c(%d)\n" i (walk i);
      loop (i + 1) n
    ) in
  loop 1 10

*)
